### ISSUE TITLE
Possible missbehaviour with KC_PROXY=edge on reverse proxy

### ISSUE THREAD

**Issue by filipetavares** [keycloak/keycloak#11807]
### Describe the bug

Following (https://github.com/keycloak/keycloak/issues/10817), I've updated keycloak the newest version (18), and I'm still running in to some issues.

As described, the current proxy setting is **edge**

First issues is that when the server is first spawn up, we get a **redirect_uri** error:
![redirect_uri](https://user-images.githubusercontent.com/3610553/166299423-efc6031d-d76f-426f-bd72-3b374a17805d.PNG)

LOGS:
> DEBUG [io.quarkus.vertx.http.runtime.ForwardedParser] (executor-thread-11) Recalculated absoluteURI to http://something.com/auth/realms/master/protocol/openid-connect/auth?client_id=security-admin-console&redirect_uri=https%3A%2F%2Fsomething.com%2Fauth%2Fadmin%2Fmaster%2Fconsole%2F&state=fbd3f10a-d929-4af8-a4ce-e3ef06c1b89e&response_mode=fragment&response_type=code&scope=openid&nonce=ea937eae-bcc6-4bf3-aae3-0a4b3742c9d7&code_challenge=pNq3qur9_XOZfrTirOaHRMxCx8pdGL-yeFjAKbLLiHA&code_challenge_method=S256

> DEBUG [org.keycloak.protocol.oidc.utils.RedirectUtils] (executor-thread-11) replacing relative valid redirect with: http://something.com/auth/admin/master/console/*

> WARN  [org.keycloak.events] (executor-thread-11) type=LOGIN_ERROR, realmId=a1634d31-f503-4b6b-9ce9-522e84855fc7, clientId=security-admin-console, userId=null, ipAddress=40.41.43.44, error=invalid_redirect_uri, redirect_uri=https://something.com/auth/admin/master/console/

Here we can see that calls are reaching the keycloak server as **http**, so I'm guessing that nginx ingress is forwarding the headers correctly, however its expected that the server knows that its being redirected from an **https** origin and should allow for access as such. This does not happen as we get the aforementioned **redirect_uri** error.

To fix this, we need to add the **absoluteURI** to the redirect_uris of the **security-admin-console** client using kcadm:
> ./kcadm.sh update clients/<ID> -s 'redirectUris=["https://something.com/auth/*"]' --no-config --server http://localhost:8080/auth --realm master --user admin --password admin

We can now enter the password and the user to get access to the admin console, but some requests are generated with the wrong **scheme**:
![keyckloak_wrong_scheme](https://user-images.githubusercontent.com/3610553/166296260-161ffb1c-87e8-4437-850f-e7e58c464f11.png)

If we copy the request as cURL we get the following:
>  curl '**http**://something.com/auth/admin/realms' -H 'User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:95.0) Gecko/20100101 Firefox/95.0' -H 'Accept: application/json, text/plain, */*' -H 'Accept-Language: en-US,en;q=0.5' -H 'Accept-Encoding: gzip, deflate' -H 'Authorization: Bearer <_TOKEN_>' -H 'Origin: https://something.com'

As we can see the address is generated with **HTTP**, at this point the browser identifies this as a possible risk and flags the requests as **MIXED BLOCK**. I'm not able to get any logs from this as the browser doesn't even try to reach the server. However if we take this and change to **HTTPS** and run the cURL command its is successful.

There are however some requests that are generated correctly:
![keyckloak_correct_scheme](https://user-images.githubusercontent.com/3610553/166297548-ae419682-c8e4-410d-80c8-abff4cab7f78.png)

So my guess is that something in the client-sided application is generating the urls incorrectly.

Is this a known bug or is it some miss configuration on my side?

My current setup is as follows:
KEYCLOAK 18 (quarkus):
  KC_PROXY=EDGE
  KC_HOSTNAME=something.com
  KC_PORT = 8080
  KC_HTTP_RELATIVE_PATH=/auth

Cloud Provicer is AWS
Running on Kubernentes (EKS)
INGRESS-CONTROLLER = nginx
Ingress variables:
    host: something.com
    path: /
    port: 8080

### Version

18

### Expected behavior

Not getting redirect_url error and being able to access the admin console.

### Actual behavior

Getting redirect_url on a fresh instance and not being able to access the admin console.

### How to Reproduce?

Spawn a new keycloak instance in kubernetes with **KC_PROXY=edge** and with an nginx ingress. Then just access the browser on the defined url.

### Anything else?

https://github.com/keycloak/keycloak/issues/10817
https://github.com/keycloak/keycloak/issues/11667
https://keycloak.discourse.group/t/mixed-block-error-on-api-request-on-the-admin-console/15178

**Comment by pedroigor** [comment]
@filipetavares Can you please confirm that the nginx is doing TLS termination?

Also, what happens if you try to access the server using `https://something.com` and then accessing the administration console?

**Comment by filipetavares** [comment]
I'm pretty sure it does, that is why we are getting **http** in the server logs, since I'm accessing the browser using **https**.

The process that I've described is accessing se application starting at `https://something.com/auth` ( I also have the relative_http_path set to /auth, forgot to mention that.)

**Comment by frozencesium** [comment]
Quarkus based Keycloak version:  18.0.0
Proxy Apache2

I'm having the exact same problem running apache2 as the proxy.  I was able to create the admin user via localhost:8080.  I have never been able to get into the master realm to configure anything.  I am not setting a relative_http_path either.  Going through the proxy I get:

 WARN  [org.keycloak.events] (executor-thread-26) type=CODE_TO_TOKEN_ERROR, realmId=e5918c51-72e4-447d-8b74-05a08711fe60, clientId=security-admin-console, userId=63d3c1be-8881-46de-8f91-9f84fa8816ff, ipAddress=(redacted), error=invalid_code, grant_type=authorization_code, code_id=7870d1b4-9cad-421e-9118-03e0814a980a, client_auth_method=client-secret

Going to the host on port 8443 directly I get:

WARN  [org.keycloak.events] (executor-thread-26) type=LOGIN_ERROR, realmId=e5918c51-72e4-447d-8b74-05a08711fe60, clientId=security-admin-console, userId=null, ipAddress=(redacted), error=invalid_redirect_uri, redirect_uri=http://example.com:8443/admin/master/console/

(obviously I'm not using example.com, just changing the hostname to protect the guilty)

All SSL certs are valid and there are a bunch of tables in the DB, so Keycloak is talking and writing to the DB.

################# Config ######################

# Database

# The username of the database user.
db-username=redacted

# The password of the database user.
db-password=redacted

# The full database JDBC URL. If not provided, a default URL is set based on the selected database vendor.
db-url=jdbc:postgresql://example.com:5432/redacted?sslmode=require

# Vault

# If set, secrets can be obtained by reading the content of files within
# the given directory
vault-dir=/var/keycloak/vault

# Logging

# Set the log file path and filename
log-file=/var/log/keycloak/keycloak.log

# Enable one or more log handlers in a comma-separated list
log=file,console

# Enable or disable colors when logging to console
log-console-color=true

# HTTP

# We're letting apache handle the ssl
http-enabled=true

# The used HTTP host
http-host=0.0.0.0

# The file path to a server certificate or certificate chain in PEM format.
https-certificate-file=/etc/ssl/certs/server.crt

# The file path to a private key in PEM format.
https-certificate-key-file=/etc/ssl/certs/server.key

# The proxy address forwarding mode if the server is behind a reverse proxy.
proxy=edge

# Hostname for the Keycloak server.
hostname=example.com

**Comment by filipetavares** [comment]
I'm still stuck on this issue, but now I'm trying to do thing with KC_PROXY=passthrough to see if it works that way, but i'm getting the exact same problem. And I found out that some links are still being generated with **http**.
![http_requets](https://user-images.githubusercontent.com/3610553/166901412-1ed817ee-57a4-4a73-a36f-eac0bac42be7.png)

Is it possible that this might still be some kind of frontend bug related with https://github.com/keycloak/keycloak/issues/10817?

**Comment by frozencesium** [comment]
Confirmed that setting proxy=reencrypt gives the same behavior.

**Comment by pedroigor** [comment]
Here is what I'm doing, please tell me if I'm missing some step.

Start a Keycloak 18 instance:

```
./kc.sh start --auto-build --health-enabled true --hostname mykeycloak --proxy edge
```

My proxy is doing TLS termination at `mykeycloak` domain.

Access `http://localhost:8080` and what I get is a `invalid_redirect_uri` error because the admin console client is using `localhost` within the redirect URI. This error is because you should now excplitly enable redirect URIs using `localhost`. The server expect it to be based on the hostname (mykeycloak).

**Comment by frozencesium** [comment]
> Here is what I'm doing, please tell me if I'm missing some step.
> 
> Start a Keycloak 18 instance:
> 
> ```
> ./kc.sh start --auto-build --health-enabled true --hostname mykeycloak --proxy edge
> ```
> 
> My proxy is doing TLS termination at `mykeycloak` domain.
> 
> Access `http://localhost:8080` and what I get is a `invalid_redirect_uri` error because the admin console client is using `localhost` within the redirect URI. This error is because you should now excplitly enable redirect URIs using `localhost`. The server expect it to be based on the hostname (mykeycloak).


In my case I don't access http://localhost:8080.  Going to https://mykeycloak doesn't work, going to http://localhost:8080 doesn't work, going to https://mykecloak:8443 doesn't work...none of it works.  The redirect_uri is setting it to http://mykeycloak, and the redirect_uri can't be changed except by digging in the DB (which I haven't done yet) because you can't log in to the admin console to change it.

**Comment by filipetavares** [comment]
I'm using nginx ingress controller in kubernetes in aws (so there is the nlb in front of it, but I'm guessing that is not the problem)... but my proxy config is roughly as follows:
```
...
server {
  server_name something.com
  listen 80;
  listen 443 proxy_protocol ssl http2;

  location / {
    proxy_set_header Host                   $host;

    proxy_set_header X-Forwarded-For        $remote_addr;

    proxy_set_header X-Forwarded-Host       $host;
    proxy_set_header X-Forwarded-Port       $port;
    proxy_set_header X-Forwarded-Proto      $scheme;
    proxy_set_header X-Forwarded-Scheme     $scheme;

    proxy_set_header X-Scheme               $scheme;

    # Pass the original X-Forwarded-For
    proxy_set_header X-Original-Forwarded-For $http_x_forwarded_for;

    # mitigate HTTPoxy Vulnerability
    # https://www.nginx.com/blog/mitigating-the-httpoxy-vulnerability-with-nginx/
    proxy_set_header Proxy                  "";
    
    proxy_next_upstream                     error timeout;
    proxy_next_upstream_timeout             0;
    proxy_next_upstream_tries               3;
    proxy_redirect                          off;
    

    proxy_pass https://<server_addr>;
  }
}
```

this is mostly auto generated by the ingress controller.

**Comment by frozencesium** [comment]
> Are you able to give your proxy configuration so that I can try to use it here and see how it goes? If you are using HAProxy even better ...

I'm using apache2 on the same VM as Keycloak, systems are hosted in AWS.

I'm using the same settings for Keycloak 14.0.0 and it works perfectly, but is broken on 18.0.0.  I am NOT migrating, this is a brand new instance of 18:

/etc/apache2/sites-enabled/default-ssl.conf

Inside the virtualhost for _default_:443:

                <Proxy *>
                        Order deny,allow
                        Allow from all
                </Proxy>
                SSLCertificateFile      /etc/ssl/certs/server.crt
                SSLCertificateKeyFile   /etc/ssl/certs/server.key

                SSLProxyEngine          On
                ProxyRequests           Off
                ProxyPreserveHost       On
                ProxyPass /auth http://localhost:8080/auth
                ProxyPassReverse /auth http://localhost:8080/auth/
                ProxyPass               /       http://localhost:8080/
                ProxyPassReverse        /       http://localhost:8080/

**Comment by pedroigor** [comment]
@filipetavares I have changed the default configuration of my nginx to match yours:

```
server {
        listen       443 ssl http2;
        listen       [::]:443 ssl http2;
        server_name  mykeycloak;
        root         /usr/share/nginx/html;

        ssl_certificate "cert.pem";
        ssl_certificate_key "key.pem";
        ssl_session_cache shared:SSL:1m;
        ssl_session_timeout  10m;
        ssl_ciphers PROFILE=SYSTEM;
        ssl_prefer_server_ciphers on;

        # Load configuration files for the default server block.
        include /etc/nginx/default.d/*.conf;

        error_page 404 /404.html;
            location = /40x.html {
        }

        error_page 500 502 503 504 /50x.html;
            location = /50x.html {
        }
        
        location / {
	    proxy_set_header Host                   $host;

    	    proxy_set_header X-Forwarded-For        $remote_addr;

    	    proxy_set_header X-Forwarded-Host       $host;
    	    proxy_set_header X-Forwarded-Port       443;
    	    proxy_set_header X-Forwarded-Proto      $scheme;
    	    proxy_set_header X-Forwarded-Scheme     $scheme;

    	    proxy_set_header X-Scheme               $scheme;

    	    # Pass the original X-Forwarded-For
    	    proxy_set_header X-Original-Forwarded-For $http_x_forwarded_for;

    	    # mitigate HTTPoxy Vulnerability
    	    # https://www.nginx.com/blog/mitigating-the-httpoxy-vulnerability-with-nginx/
    	    proxy_set_header Proxy                  "";
    
    	    proxy_next_upstream                     error timeout;
    	    proxy_next_upstream_timeout             0;
    	    proxy_next_upstream_tries               3;
    	    proxy_redirect                          off;
    

    	    proxy_pass http://localhost:8080;
  	}
    }
```

I had to change the following setting because the variable `port` was failing to resolve:

```
proxy_set_header X-Forwarded-Port       443;
```

The server is started as follows:

```
./kc.sh start --auto-build --health-enabled true --hostname mykeycloak --proxy edge
```

A request to `https:/mykeycloak` and then to the admin console by clicking on the `Administration Console` when at the welcome page redirect me successfuly to the login page.

**Comment by pedroigor** [comment]
@frozencesium When you mention 18.0.0 I guess you are referring to the new distribution. Did you try to run using 18.0.0-legacy (wildfly)?

I'm wondering if it can be related to both distributions somehow.

**Comment by frozencesium** [comment]
@pedroigor I did not try legacy wildfly because it's, by Keycloak's own docs...legacy.  I can try, but that isn't going to fix the problem if there is a bug in Quarkus distro.

**Comment by pedroigor** [comment]
Yeah, trying to see if it is strict to Quarkus because we changed the default hostname provider.

I'm going to check your Apache configuration too and see if I can reproduce it and finally see what might be happening.

**Comment by frozencesium** [comment]
@pedroigor switching to nginx with the config you posted worked.  Something changed between Keycloak 14 and 18 to cause that simple apache2 config that I posted earlier to fail.  I would be interested to see what other options need to be set in apache2 to get the proxy to work as expected.

**Comment by pedroigor** [comment]
@frozencesium This one works for me:

```
<VirtualHost *:443>
        ServerName keycloak
        ServerAdmin webmaster@localhost
        DocumentRoot /var/www/html

        # Sample for Keycloak
        ProxyPass /auth http://127.0.0.1:8080/auth
        ProxyPassReverse /auth http://127.0.0.1:8080/auth

        #Can ignore SSL Check on destination proxy example auto signed certificate on springboot
        SSLProxyCheckPeerCN off
        SSLProxyCheckPeerExpire off
        SSLProxyCheckPeerName off
        SSLProxyEngine On
        SSLProxyVerify none

        ProxyPreserveHost On
        ProxyPass /secure https://127.0.0.1:8443/special_secure
        ProxyPassReverse /secure https://127.0.0.1:8443/special_secure

        RequestHeader set X-Forwarded-Proto https
        # Let's encrypt
        SSLCertificateFile conf.d/server.crt
        SSLCertificateKeyFile conf.d/server.key

</VirtualHost>
```

**Comment by pedroigor** [comment]
@DGuhr Perhaps the discussion here can be useful when you are doing the repository with proxy examples.

**Comment by manodupont** [comment]
> server {
>         listen       443 ssl http2;
>         listen       [::]:443 ssl http2;
>         server_name  mykeycloak;
>         root         /usr/share/nginx/html;
> 
>         ssl_certificate "cert.pem";
>         ssl_certificate_key "key.pem";
>         ssl_session_cache shared:SSL:1m;
>         ssl_session_timeout  10m;
>         ssl_ciphers PROFILE=SYSTEM;
>         ssl_prefer_server_ciphers on;
> 
>         # Load configuration files for the default server block.
>         include /etc/nginx/default.d/*.conf;
> 
>         error_page 404 /404.html;
>             location = /40x.html {
>         }
> 
>         error_page 500 502 503 504 /50x.html;
>             location = /50x.html {
>         }
>         
>         location / {
> 	    proxy_set_header Host                   $host;
> 
>     	    proxy_set_header X-Forwarded-For        $remote_addr;
> 
>     	    proxy_set_header X-Forwarded-Host       $host;
>     	    proxy_set_header X-Forwarded-Port       443;
>     	    proxy_set_header X-Forwarded-Proto      $scheme;
>     	    proxy_set_header X-Forwarded-Scheme     $scheme;
> 
>     	    proxy_set_header X-Scheme               $scheme;
> 
>     	    # Pass the original X-Forwarded-For
>     	    proxy_set_header X-Original-Forwarded-For $http_x_forwarded_for;
> 
>     	    # mitigate HTTPoxy Vulnerability
>     	    # https://www.nginx.com/blog/mitigating-the-httpoxy-vulnerability-with-nginx/
>     	    proxy_set_header Proxy                  "";
>     
>     	    proxy_next_upstream                     error timeout;
>     	    proxy_next_upstream_timeout             0;
>     	    proxy_next_upstream_tries               3;
>     	    proxy_redirect                          off;
>     
> 
>     	    proxy_pass http://localhost:8080;
>   	}
>     }

I do not understand that!

Why do you need to set certificates, if you start with --proxy=edge ?

**Comment by DGuhr** [comment]
@manodupont see https://github.com/keycloak/keycloak/discussions/11856, perhaps this provides some more context, especially the part around "The Issue with HOSTNAME_STRICT_HTTPS".

**Comment by manodupont** [comment]
@pedroigor  Oh no sorry! I was misleading something. Its the front facing certificates that allows client to make https request. It is indeed proxying to http afterwards. Ok i will try this with the real config. My problem is that I tried that, but Im uysing Kubernetes ingresses sooo i think some nginx conf might not work properly using the annotations.

I will try to use this exact config

**Comment by manodupont** [comment]
@DGuhr I ve read your doc. Its pretty clear. But what about that redirect_uri error that @filipetavares mentionned in this ticket ? Is it related to only change ? is it supposed to work with `proxy=edge` and `HOSTNAME_STRICT_HTTPS=false` ?

**Comment by DGuhr** [comment]
@manodupont You'd need to set `HOSTNAME_STRICT_HTTPS=False` to allow HTTP _in general_, yes. **But**: the exact issue requires another step: 

The problem is due to the fact that the actual `redirect_uri` for the `security-admin-console` client in the `master` realm is using a `valid redirect uris` entry that is relative to the hostname you set for keycloak. See picture below.
![Screenshot 2022-05-06 at 10 49 24](https://user-images.githubusercontent.com/89905860/167100705-1c8b7dcf-7dd5-46eb-9cfa-f933f6a11d0a.png)

e.g. you have keycloak running on `https://id.keycloak.test:8443` with ` start --proxy=edge` (or `--http-enabled=true`) plus  `--hostname-strict-https=false` and you want to access it internally using `http:localhost:8080`. So you click on `Administration console` on the welcome page. You'll see you are redirected to `id.keycloak.test:8443/...`when the login form shows up. After adding your credentials and submitting them, you see:
![Screenshot 2022-05-06 at 10 48 42](https://user-images.githubusercontent.com/89905860/167101880-d4e83a0f-3f38-4ad5-b173-d32d22a25178.png)

This is technically expected atm, as the configuration for the console, as stated above, looks like this ootb:
![Screenshot 2022-05-06 at 10 48 42](https://user-images.githubusercontent.com/89905860/167101103-bb531e8e-4f87-456c-b10e-5de293380f4f.png)

When you add the correct `redirect_uri` to the valid redirect uris entries, it'll work. You could e.g. use the [kcadm.sh admin client ](https://www.keycloak.org/docs/latest/server_admin/index.html#installing-the-admin-cli) to do that automated.

So when you have set it like in the screenshot below:

![Screenshot 2022-05-06 at 11 07 45](https://user-images.githubusercontent.com/89905860/167102555-0b4172ee-102c-47d5-94ce-b889a6eedc45.png)

the click on submit here:
![Screenshot 2022-05-06 at 10 50 17](https://user-images.githubusercontent.com/89905860/167102652-7aab2582-828f-4c3b-b0a4-881eac29fc4c.png)

will work then. (see the url I access the admin console on the picture above for proof).

Hope this makes it clearer?

**Comment by manodupont** [comment]
@DGuhr Ok. So as i understand, the only step I missed, is to rightly setup the redirect_uri for the console. Ok fine.
But do we agree that this is a missing behaviour that can be fix directly in the code ? having to set this up manually by using a script seems weird, right. I don't remember having to do that in some other version!

**Comment by pedroigor** [comment]
@filipetavares Yes, it is. You should not face those errors.

**Comment by manodupont** [comment]
After one afternoon, honestly, I confirm something is not right. I don't know .... I have tried all possible solutions, nothing works when using production for me. Even the full https end-to-end on https://0.0.0.0:8443 keycloak ... still I get this redirect_uri error.

**Comment by filipetavares** [comment]
You do still need to update the redirect_uri manual as far as I know... Can you try it, and check you can access the admin console? Just to confirm that you don't get the same error that I do.

**Comment by manodupont** [comment]
i do. i do have all mixed content afterwards.

**Comment by manodupont** [comment]
And on top of that. i do want to use passthrough or reencrypt and with those settings, i cannot get through the admin console. i dont want to use edge

**Comment by filipetavares** [comment]
I tried with passthrough as well... but I'm getting the exact same result

**Comment by pedroigor** [comment]
For sure, something is not well in Keycloak.

Can you try removing the `KC_HOSTNAME_STRICT_HTTPS` option? You shouldn't need it in this case because you don't want Keycloak accessed through HTTP so all URLs should be built based on the HTTP scheme. That might explain the CSP error in the browser.

**Comment by filipetavares** [comment]
I don't use that envar, didn't even know it existed to be honest.

**Comment by pedroigor** [comment]
W.r.t. to the redirect uri error. In this case, you should not get that. I'll try some configurations here and also run k8s (although I think that my local env should work).

**Comment by manodupont** [comment]
In comparison, i found an old project of mine, which was using version 11 back then, and the only option `PROXY_ADDRESS_FORWARDING=true` was giving me https to the console link

<img width="1536" alt="Screenshot 2022-05-07 at 10 16 26" src="https://user-images.githubusercontent.com/7568033/167245673-e0542986-7d87-4654-ba19-c45a1345feaa.png">

**Comment by FritzJo** [comment]
Any update on this? The OIDC login for individual services works perfectly, but I can't access the admin console. For some reason the redirect URL is invalid since the last update and "Back to Application" takes me to ```https://kc.example.com:80/auth/admin/master/console/```. I'm not sure where the port ```80``` comes from, because neither ```KC_HTTP_PORT```, nor ```KC_HTTPS_PORT``` are set to ```80```. 
I'm running KC with Docker behind a Nginx reverse proxy with ```KC_PROXY=edge```. 
Rolling back to version 17 results in a blank page like described in https://github.com/keycloak/keycloak/issues/10817

**Comment by fwhenin** [comment]
I'm having something similar that I brought up in [discussions the other day](https://github.com/keycloak/keycloak/discussions/12028). Except for some reason it wants 0.0.0.0 as the assumed host. This is running on ECS Fargate with an Application Load Balancer and a custom domain in Route53 pointing to the Application Load Balancer. My environment variables are as follow

        ```
KC_PROXY: "edge",
        KC_HOSTNAME: `login.domain.com`,
        KC_HOSTNAME_STRICT: "false",
        KC_HOSTNAME_STRICT_HTTPS: "false",
        KC_HOSTNAME_STRICT_BACKCHANNEL: "true",
        KC_LOG_LEVEL: "DEBUG",
```

**Comment by fwhenin** [comment]
I just used the 17.0.0 version of keycloak, and I'm not seeing the issue anymore

**Comment by fwhenin** [comment]
another update: I just did the same exact deployment with 17.0.1, and the issue came up again. Reverting back to 17.0.0 without clearing the database didn't make the issue go away. Could this be something that sticks around in the database after going to 17.0.1?

**Comment by pedroigor** [comment]
@fwhenin Not sure, did you try v18?

One thing you can try is to make sure your proxy is forwarding the expected host name (`login.domain.com`).

**Comment by filipetavares** [comment]
So I've tried to replicate the same thing with a local kubernetes, and I don't seem to be able to reproduce the error. After some digging, I've figured that the problem is with the AWS Network LoadBalancer (which I also very lightly mention on this issue) that does TLS Termination. 

I've messed a bit with it, and if remove the certificate from the NLB it will stop the TLS Termination at the NLB level, but I am not able to use the certificate in the Nginx-Controllers because, as far as I know, I can't use ACM (Amazon Certificate Manager) certificates in the eks workers (or atleast not if I don't install an extra costly plugin which is Nitro-Enclave, which also makes me recreate my machines since it requires a new image for it to be supported).

This however doesn't explain why some requests are still being return as "https" while others are returned as "http", which might mean that some requests are being treated differently.

**Comment by pedroigor** [comment]
@filipetavares Admin Urls are treated differently because the schema is resolved from the request itself.

When using `proxy=edge` where your instances are listening plain HTTP, you should set the `X-Forwarded-Proto` headers so that the HTTPS schema is properly set to the URLs created by the server.

W.r.t. to the `invalid_redirect_uri` error, it should be fixed by https://github.com/keycloak/keycloak/pull/12192.

**Comment by filipetavares** [comment]
The `X-Forwarded-Proto` seems to be set correctly at the nginx level, however since the NLB is offloading, the nginx is receiving the request as plain http. 

From reading the ingress-controller documentation I also believe that this might be true for Azure (although I haven't tested it).

**Comment by manodupont** [comment]
Is this really closed ? I didn't see any fix or solution ?

**Comment by pedroigor** [comment]
@manodupont Sorry for not giving more context. The original issue about the invalid redirect_uri error should be fixed now we re-introduced the `hostname-admin` option.

See https://github.com/keycloak/keycloak/pull/12192.

**Comment by manodupont** [comment]
Ok. I'll check that. Thanks a lot

**Comment by filipetavares** [comment]
That was part of the original issue. But I've come to understand that the problem lies in the way AWS sets up its infrastructure, since we have two load balancers in front of the application (the entrance to the cloud/VPC which is the network load balancer, NLB, and the kubernetes nginx ingress), where the NLB terminates TLS, which apparently is the recommended behavior.

Right now I've shifted priorities, but when I come back to this issue again I'll open another github issue with more details.

Anyways, thanks for the help, the fix and keep up the good work.

**Comment by andewil** [comment]
Just for sharing: I solve the same problem with redirect_uri by adding in table redirect_uris an additional URI for domain. In my case it was https://<initial_domain>/* for client "security-admin-console". ID could be extracted from "client" table.

**Comment by hellobiek** [comment]
> That was part of the original issue. But I've come to understand that the problem lies in the way AWS sets up its infrastructure, since we have two load balancers in front of the application (the entrance to the cloud/VPC which is the network load balancer, NLB, and the kubernetes nginx ingress), where the NLB terminates TLS, which apparently is the recommended behavior.
> 
> Right now I've shifted priorities, but when I come back to this issue again I'll open another github issue with more details.
> 
> Anyways, thanks for the help, the fix and keep up the good work.

my condition is same with you，have you fixed this problem?

**Comment by filipetavares** [comment]
@hellobiek I haven't looked much in to it afterwards, but at the time, what I did was to remove the certificate from the NLB, which disables the TLS termination at the NLB level, and create a certificate to the nginx ingress and have it terminate TLS there. If I remember correctly that fixed the issue, although it means that you have to manage certificates in your nginx instead of using the aws certificate manager ( you can still use the manager with the nginx, but you need to install a couple of plugins [https://cert-manager.io/ , https://github.com/cert-manager/aws-privateca-issuer] and its not that straight forward).

**Comment by Camsteack** [comment]
So at the moment there is no way to run Keycloak X with an NLB terminating TLS and nginx ingress controller behind ? I have been trying to fix that for a few days now. It was working with with Keycloak legacy (wildfly). I really don't want to have to manage certificates outside of the AWS certificate manager.
Isn't there any other workaround ? This setup seems like a pretty classic setup these days
Many thanks !

**Comment by pedroigor** [comment]
@Camsteack I don't see why it should not work. When using proxy `edge` you are basically forwarding requests through HTTP. You just need to make sure the hostname settings are pointing to your NLB accordingly and using `https` and the expected domain name.

**Comment by Camsteack** [comment]
Hey @pedroigor, I am not sure why it does not work but I have the same issues as the last few posts. The admin console is not available as the redirect uri is invalid and if I change it manually in the DB I then get mixed content when loading the console. The hostname settings and protocol are setup correctly and I have tried every combination possible. The same configuration works fine with Keycloak Legacy. I can still use keycloak as login to custom realms and clients is working but I cannot access the admin console.
It seems that everyone with the same setup (NLB terminating TLS -> nginx ingress controller -> keycloak) is still facing the same issue 🤷

**Comment by manodupont** [comment]
+1

**Comment by pedroigor** [comment]
If the admin console is failing to load due to the invalid redirect URI error it is probably because you are trying to access it using a host other than you set to `hostname`. If you want to access the admin console using a different URL, you can set the `hostname-admin` option to set a different host name. Did you try this option already?

Btw, as I mentioned before, we are about to introduce two additional options to set the full base URL for both "frontend" and "admin" URLs similarly to the legacy distribution.

**Comment by Camsteack** [comment]
Hi @pedroigor thanks for the quick replies. I am not accessing it using a different hostname, it is the same one everywhere. The issue is also not the hostname but the protocol as like stated in some posts above, if I change the redirect_uri parameter in the login URL from `https` to `http` then it works, so the hostname is the correct one, just the protocol is not the one keycloak expects.

**Comment by pedroigor** [comment]
I see. Can you check if the `X-Forwarded-Proto` is being set?

If you are not able to configure your proxy to set this header then you are blocked by https://github.com/keycloak/keycloak/issues/13364#issuecomment-1204409513. The PR solving the problem is https://github.com/keycloak/keycloak/pull/13681.

I think now is more clear what is happening.

**Comment by Camsteack** [comment]
I think you are correct, that might fix it and it should work as it was working in the legacy distribution !

So the nginx proxy is forwarding those headers but because we are using a Layer 4 NLB they are not set there so I think that what is causing the issue. 
Trying to hardcode the `X-Forwarded-Proto` with the annotations doesn't work either
```
      nginx.ingress.kubernetes.io/configuration-snippet: |
        proxy_set_header X-Forwarded-For $host;
        proxy_set_header X-Forwarded-Proto https
```
Do you know when https://github.com/keycloak/keycloak/pull/13681 is expected to be released ? :)
Thanks !

**Comment by pedroigor** [comment]
I would like to have it in 19.0.2. Yet to confirm. But it should be in v20 for sure.

I do expect the server to respect the scheme set via `X-Forwarded-Header` we do have tests covering this behavior. Perhaps you can look at the access logs and see the headers from requests by using this configuration from Quarkus. See https://quarkus.io/guides/http-reference#configuring-http-access-logs and [Using Raw Quarkus Properties](https://www.keycloak.org/server/configuration).

**Comment by Camsteack** [comment]
I have just tested 19.0.2 and it works now ! Thanks a lot @pedroigor :)

**Comment by 0xdreadnaught** [comment]
So I think I figured out what's going on here.
I was trying to deploy Keycloak in a project that has all of the containers 100% reliant on the docker network for communication with the exception of a single nginx reverse proxy that had an http and https port exposed. As such I didn't put a `ports:` section in my compose. 

As soon as I added the following for testing, the port appears in the admin URL link like it should.
```
ports:
 - 8088:8080
 - 8888:8443
 ```
 
 Creates a new problem for me, hopefully solves yours. This could be fixed with an environment variable if it can be configured to search for them.
